src/GafferScene/InteractiveRender.cpp

```diff
@@ -57,6 +61,71 @@ using namespace GafferScene;
 
 IE_CORE_DEFINERUNTIMETYPED( InteractiveRender );
 
+//////////////////////////////////////////////////////////////////////////
+// SceneGraph implementation
+//
+// This is a node in a scene hierarchy, which gets built when the
+// interactive render starts up. Each node stores the attribute hash of
+// the input scene at its corresponding location in the hierarchy, so
+// when the incoming scene updates, we are able to determine the locations
+// at which the attributes have changed since the last update, reevaluate
+// those attributes and send updates to the renderer.
+//
+// \todo: This is very similar to the SceneGraph mechanism in
+// GafferSceneUI::SceneGadget. At some point it would be good to refactor
+// this and use the same class for both of them. See the comments in
+// src/GafferSceneUI/SceneGadget.cpp for details.
+//
+//////////////////////////////////////////////////////////////////////////
+
+class InteractiveRender::SceneGraph
+{
+
+	public :
+
+		SceneGraph() : m_parent(0)
+		{
+		}
+
+		~SceneGraph()
+		{
+			for( std::vector<SceneGraph *>::const_iterator it = m_children.begin(), eIt = m_children.end(); it != eIt; ++it )
+			{
+				delete *it;
+			}
+		}
+		
+		void path( ScenePlug::ScenePath &p )
+		{
+			if( !m_parent )
+			{
+				return;
+			}
+			m_parent->path( p );
+			p.push_back( m_name );
+		}
+
+	private :
+		
+		friend class SceneGraphBuildTask;
+		friend class SceneGraphIteratorFilter;
+		friend class SceneGraphEvaluatorFilter;
+		friend class SceneGraphOutputFilter;
+		
+		// scene structure data:
+		IECore::InternedString m_name;
+		SceneGraph *m_parent;
+		std::vector<InteractiveRender::SceneGraph *> m_children;
+		
+		// hash of the attributes as of the most recent evaluation:
+		IECore::MurmurHash m_attributesHash;
+		
+		// actual scene data:
+		IECore::ConstCompoundObjectPtr m_attributes;
+		IECore::ConstObjectPtr m_object;
+		IECore::M44fDataPtr m_transform;
```

Comment:
I notice that lower down, you have to `new M44fData` to put in here, because `transformPlug()`->getValue()` just returns a plain `M44f`. Are you using a pointer so you can use null-ness as a "nothing to do" sentinel value? Maybe `boost::optional` would be another possibility? From my understanding, boost::optional doesn't need to allocate memory. Another possibility would be to allow `M44fPlug` (or all ValuePlugs) to return `Ptr` as well, although I was trying to hide that implementation detail.